service/domain/replication/ebt/replicator.go

```diff
@@ -0,0 +1,111 @@
+package ebt
+
+import (
+	"context"
+	"time"
+
+	"github.com/boreq/errors"
+	"github.com/planetary-social/scuttlego/logging"
+	"github.com/planetary-social/scuttlego/service/domain/messages"
+	"github.com/planetary-social/scuttlego/service/domain/transport"
+	"github.com/planetary-social/scuttlego/service/domain/transport/rpc"
+)
+
+const (
+	waitForRemoteToStartEbtSessionFor = 5 * time.Second
+	ebtReplicateVersion               = 3
+)
+
+var (
+	ebtReplicateFormat = messages.EbtReplicateFormatClassic
+)
+
+type Replicator struct {
+	tracker *SessionTracker
+	runner  *SessionRunner
+	logger  logging.Logger
+}
+
+func NewReplicator(tracker *SessionTracker, runner *SessionRunner, logger logging.Logger) Replicator {
+	return Replicator{
+		tracker: tracker,
+		runner:  runner,
+		logger:  logger,
+	}
+}
+
+func (r Replicator) Replicate(ctx context.Context, peer transport.Peer) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	connectionId, ok := rpc.GetConnectionIdFromContext(ctx)
+	if !ok {
+		return errors.New("connection id not found in context")
+	}
+
+	logger := r.logger.WithField("peer", peer)
+
+	if !peer.Conn().WasInitiatedByRemote() {
+		logger.Debug("initializing an EBT session")
+
+		done, err := r.tracker.OpenSession(connectionId)
```

Comment:
Test: shouldn't this wait for session?

Reply:
Maybe not? If someone opened a session when they shouldn't that this implies that it is an error? I just documented this in tests and we can change this in the future.